File: Python/day23/AOC.py
def part1(input, part2=False):
    registers = {"a": 0, "b": 0}
    if part2:
        registers["a"] = 1
    i = 0
    while i < len(input):
        inst = input[i].split()
        match inst[0]:
            case "hlf":
                registers[inst[1]] /= 2
                i += 1
            case "tpl":
                registers[inst[1]] *= 3
                i += 1
            case "inc":
                registers[inst[1]] += 1
                i += 1
            case "jmp":
                i += int(inst[1])
                continue
            case "jie":
                if registers[inst[1][0]] % 2 == 0:
                    i += int(inst[2])
                    continue
                else:
                    i += 1
            case "jio":
                if registers[inst[1][0]] == 1:
                    i += int(inst[2])
                    continue
                else:
                    i += 1
    return registers

def part2(input):
    return part1(input, part2=True)

def main():
    input = open("input.txt", "r").read().splitlines()
    part1_ans = part1(input)
    print(f"Part 1: {part1_ans['b']}")
    part2_ans = part2(input)
    print(f"Part 2: {part2_ans['b']}")

if __name__ == "__main__":
    main()

File: Go/day09/AOC.go
package main

import (
	"fmt"
	"os"
	"slices"
	"strings"
)

type location string
type distance int

type node struct {
	name      location
	neighbors map[location]distance
}

func (n *node) addNeighbor(loc location, dist distance) {
	if n.neighbors == nil {
		n.neighbors = make(map[location]distance)
	}
	n.neighbors[loc] = dist
}

func parseInput(input []string) map[location]*node {
	nodes := make(map[location]*node)

	for _, line := range input {
		var loc1, loc2 location
		var dist distance
		fmt.Sscanf(line, "%s to %s = %d", &loc1, &loc2, &dist)

		if _, ok := nodes[loc1]; !ok {
			nodes[loc1] = &node{name: loc1}
		}
		if _, ok := nodes[loc2]; !ok {
			nodes[loc2] = &node{name: loc2}
		}

		nodes[loc1].addNeighbor(loc2, dist)
		nodes[loc2].addNeighbor(loc1, dist)
	}

	return nodes
}

func generatePermutations(elements []string) [][]string {
	var result [][]string

	var generate func(int)
	generate = func(n int) {
		if n == 1 {
			temp := make([]string, len(elements))
			copy(temp, elements)
			result = append(result, temp)
			return
		}

		for i := 0; i < n; i++ {
			elements[i], elements[n-1] = elements[n-1], elements[i]
			generate(n - 1)
			elements[i], elements[n-1] = elements[n-1], elements[i]
		}
	}

	generate(len(elements))
	return result
}

func generateRoutes(nodes map[location]*node) []int {
	var routes []int
	locs := []string{}
	for loc := range nodes {
		locs = append(locs, string(loc))
	}

	perms := generatePermutations(locs)
	for _, perm := range perms {
		dist := 0
		for i := 0; i < len(perm)-1; i++ {
			dist += int(nodes[location(perm[i])].neighbors[location(perm[i+1])])
		}
		routes = append(routes, dist)
	}

	return routes
}

func part1(input []int) int {
	slices.Sort(input)
	return input[0]
}

func part2(input []int) int {
	slices.Sort(input)
	return input[len(input)-1]
}

func main() {
	input, err := os.ReadFile("input.txt")
	if err != nil {
		fmt.Print("Error reading file\n")
		panic(err)
	}

	lines := strings.Split(string(input), "\n")
	routes := generateRoutes(parseInput(lines))

	part1_ans := part1(routes)
	fmt.Printf("Part 1: %d\n", part1_ans)

	part2_ans := part2(routes)
	fmt.Printf("Part 2: %d\n", part2_ans)
}
